packages/vite/src/node/plugins/importAnalysis.ts

```diff
@@ -122,6 +122,10 @@ export function importAnalysisPlugin(config: ResolvedConfig): Plugin {
     },
 
     async transform(source, importer, options) {
+      if (server === null) {
```

Comment:
The server will always be defined here, did you have an issue that needed this guard when doing the PR?

Reply:
Yes, in a real app, it is always defined. But the test [`packages/vite/src/node/server/__tests__/pluginContainer.spec.ts`](https://github.com/vitejs/vite/blob/a74bd7ba9947be193bccf636b8918bd1f59e89ae/packages/vite/src/node/server/__tests__/pluginContainer.spec.ts) creates a plugin container and calls `transform` without creating a dev server: https://github.com/vitejs/vite/blob/a74bd7ba9947be193bccf636b8918bd1f59e89ae/packages/vite/src/node/server/__tests__/pluginContainer.spec.ts#L64.

Reply:
Could we test in somw other way, I dont think we should add these guards for the tests

Reply:
That was my first thought as well, but the test is about the plugin container. So it makes sense that it uses the plugin container without creating a dev server. We could make the test create a dummy dev server but IMO it would degrade the do-one-thing-do-it-well property of the test.

Reply:
Almost all of Vite's CI is e2e, so don't worry about having the server in the middle. Couldn't the test be added following other HMR tests in that playground? Why would this one be different?

Reply:
It's unrelated to the new HMR test.

The new `if (server === null) { return null }` is needed only because the PR moves up `const { moduleGraph } = server`. And, at that new position, `server` is not guaranteed to be defined anymore, because of the `pluginContainer.spec.ts` test.

We could modify `pluginContainer.spec.ts` to use `createServer()` instead of `createPluginContainer()` &mdash; ensuring that a dev server exists. But I'm not familiar with  `pluginContainer.spec.ts` so I'm not sure here.

Reply:
Oh, I see. Ok, I think you are right here.